maven-plugin/src/main/java/org/sonatype/nexus/maven/staging/StagingUploadMojo.java

```diff
@@ -0,0 +1,408 @@
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.maven.staging;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+import com.sonatype.nexus.api.exception.RepositoryManagerException;
+import com.sonatype.nexus.api.repository.v3.DefaultAsset;
+import com.sonatype.nexus.api.repository.v3.DefaultComponent;
+import com.sonatype.nexus.api.repository.v3.RepositoryManagerV3Client;
+import com.sonatype.nexus.api.repository.v3.Tag;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.DefaultArtifact;
+import org.apache.maven.artifact.handler.DefaultArtifactHandler;
+import org.apache.maven.artifact.repository.ArtifactRepository;
+import org.apache.maven.artifact.repository.ArtifactRepositoryFactory;
+import org.apache.maven.artifact.repository.layout.ArtifactRepositoryLayout;
+import org.apache.maven.artifact.repository.metadata.GroupRepositoryMetadata;
+import org.apache.maven.artifact.versioning.VersionRange;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.plugins.annotations.Component;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.project.artifact.ProjectArtifactMetadata;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Goal to tag and upload artifacts to NXRM 3 from a previously deferred deploy (locally staged).
+ */
+@Mojo(name = "upload", requiresProject = false, requiresOnline = true,
+    threadSafe = true)
+public class StagingUploadMojo
+    extends StagingMojo
+{
+  private static final String FORMAT = "maven2";
+
+  @Parameter(property = "repository")
```

Comment:
Should this parameter be a required parameter for upload?

Reply:
Done in [ead71ff](https://github.com/sonatype/nxrm3-maven-plugin/pull/36/commits/ead71ffee2f92e843ac341fe6d8de9bf3cbce424)